File: frontend/src/utils/permissions.ts
import { UserRole } from '../types';

// ============================================================================
// Module Permission Definitions
// ============================================================================

// 模块代码定义
export const ModuleCode = {
  WORK_ORDERS: 'work_orders',
  DASHBOARD: 'dashboard',
  LOCATIONS: 'locations',
  PERSONNEL: 'personnel',
  EQUIPMENT: 'equipment',
  METHODS: 'methods',
  MATERIALS: 'materials',
  CLIENTS: 'clients',
  PRODUCTS: 'products',
  HANDOVERS: 'handovers',
  AUDIT_LOGS: 'audit_logs',
  USER_MANAGEMENT: 'user_management',
  SETTINGS: 'settings',
} as const;

export type ModuleCode = typeof ModuleCode[keyof typeof ModuleCode];

// 路由到模块代码的映射
export const routeToModule: Record<string, ModuleCode> = {
  '/dashboard': ModuleCode.DASHBOARD,
  '/equipment-dashboard': ModuleCode.DASHBOARD,
  '/personnel-dashboard': ModuleCode.DASHBOARD,
  '/locations': ModuleCode.LOCATIONS,
  '/sites': ModuleCode.LOCATIONS,
  '/laboratories': ModuleCode.LOCATIONS,
  '/personnel': ModuleCode.PERSONNEL,
  '/skills': ModuleCode.PERSONNEL,
  '/skills-config': ModuleCode.PERSONNEL,
  '/transfers': ModuleCode.PERSONNEL,
  '/shifts': ModuleCode.PERSONNEL,
  '/equipment': ModuleCode.EQUIPMENT,
  '/methods': ModuleCode.METHODS,
  '/materials': ModuleCode.MATERIALS,
  '/work-orders': ModuleCode.WORK_ORDERS,
  '/work-order-query': ModuleCode.WORK_ORDERS,
  '/clients': ModuleCode.CLIENTS,
  '/client-slas': ModuleCode.CLIENTS,
  '/source-categories': ModuleCode.CLIENTS,
  '/products': ModuleCode.PRODUCTS,
  '/handovers': ModuleCode.HANDOVERS,
  '/audit-logs': ModuleCode.AUDIT_LOGS,
  '/user-management': ModuleCode.USER_MANAGEMENT,
  '/settings': ModuleCode.SETTINGS,
};

// 模块标签
export const moduleLabels: Record<ModuleCode, string> = {
  [ModuleCode.WORK_ORDERS]: '工单管理',
  [ModuleCode.DASHBOARD]: '仪表板',
  [ModuleCode.LOCATIONS]: '位置管理',
  [ModuleCode.PERSONNEL]: '人员管理',
  [ModuleCode.EQUIPMENT]: '设备管理',
  [ModuleCode.METHODS]: '方法管理',
  [ModuleCode.MATERIALS]: '材料管理',
  [ModuleCode.CLIENTS]: '客户管理',
  [ModuleCode.PRODUCTS]: '产品管理',
  [ModuleCode.HANDOVERS]: '交接管理',
  [ModuleCode.AUDIT_LOGS]: '审计日志',
  [ModuleCode.USER_MANAGEMENT]: '用户管理',
  [ModuleCode.SETTINGS]: '系统设置',
};

// ============================================================================
// Operation Permission Definitions
// ============================================================================

// Permission definitions based on 权限矩阵 (Permission Matrix)
export const Permission = {
  // User management
  MANAGE_USERS: 'manage_users',
  MANAGE_SITES: 'manage_sites',
  MANAGE_LABORATORIES: 'manage_laboratories',
  MANAGE_LOCATIONS: 'manage_locations',
  
  // Client & SLA
  MANAGE_CLIENTS: 'manage_clients',
  MANAGE_CLIENT_SLA: 'manage_client_sla',
  MANAGE_SOURCE_CATEGORIES: 'manage_source_categories',
  
  // Skills
  MANAGE_SKILL_CATEGORIES: 'manage_skill_categories',
  MANAGE_SKILLS: 'manage_skills',
  ASSIGN_PERSONNEL_SKILLS: 'assign_personnel_skills',
  
  // Personnel
  INITIATE_BORROW: 'initiate_borrow',
  APPROVE_BORROW: 'approve_borrow',
  MANAGE_SHIFTS: 'manage_shifts',
  
  // Equipment & Methods
  MANAGE_EQUIPMENT: 'manage_equipment',
  MANAGE_METHODS: 'manage_methods',
  
  // Work Orders
  CREATE_WORK_ORDER: 'create_work_order',
  ASSIGN_LEAD_ENGINEER: 'assign_lead_engineer',
  CREATE_SUBTASK: 'create_subtask',
  ASSIGN_TECHNICIAN: 'assign_technician',
  EXECUTE_SUBTASK: 'execute_subtask',
  VERIFY_RESULTS: 'verify_results',
  VIEW_WORK_ORDER_QUERY: 'view_work_order_query', // Viewer-specific permission
  
  // Handover
  INITIATE_HANDOVER: 'initiate_handover',
  
  // Materials
  MANAGE_MATERIALS: 'manage_materials',
  ALLOCATE_MATERIALS: 'allocate_materials',
  HANDLE_MATERIAL_RETURN: 'handle_material_return',
  CONFIRM_MATERIAL_ALERTS: 'confirm_material_alerts',
  
  // Dashboards & Reports
  VIEW_LAB_DASHBOARD: 'view_lab_dashboard',
  VIEW_ALL_DASHBOARDS: 'view_all_dashboards',
  VIEW_CYCLE_TIME_REPORT: 'view_cycle_time_report',
  VIEW_SKILLS_MATRIX: 'view_skills_matrix',
  VIEW_REPORTS: 'view_reports',
  
  // Audit
  VIEW_AUDIT_LOGS: 'view_audit_logs',
} as const;

export type Permission = typeof Permission[keyof typeof Permission];

// Role to permissions mapping based on permission matrix
const rolePermissions: Record<UserRole, Permission[]> = {
  admin: Object.values(Permission), // Admin has all permissions
  
  manager: [
    Permission.MANAGE_LOCATIONS,
    Permission.MANAGE_CLIENTS,
    Permission.MANAGE_CLIENT_SLA,
    Permission.MANAGE_SOURCE_CATEGORIES,
    Permission.MANAGE_SKILL_CATEGORIES,
    Permission.MANAGE_SKILLS,
    Permission.ASSIGN_PERSONNEL_SKILLS,
    Permission.INITIATE_BORROW,
    Permission.APPROVE_BORROW,
    Permission.MANAGE_SHIFTS,
    Permission.MANAGE_EQUIPMENT,
    Permission.MANAGE_METHODS,
    Permission.CREATE_WORK_ORDER,
    Permission.ASSIGN_LEAD_ENGINEER,
    Permission.CREATE_SUBTASK,
    Permission.ASSIGN_TECHNICIAN,
    Permission.EXECUTE_SUBTASK,
    Permission.VERIFY_RESULTS,
    Permission.VIEW_WORK_ORDER_QUERY,
    Permission.INITIATE_HANDOVER,
    Permission.MANAGE_MATERIALS,
    Permission.ALLOCATE_MATERIALS,
    Permission.HANDLE_MATERIAL_RETURN,
    Permission.CONFIRM_MATERIAL_ALERTS,
    Permission.VIEW_LAB_DASHBOARD,
    Permission.VIEW_ALL_DASHBOARDS,
    Permission.VIEW_CYCLE_TIME_REPORT,
    Permission.VIEW_SKILLS_MATRIX,
    Permission.VIEW_REPORTS,
    Permission.VIEW_AUDIT_LOGS,
  ],
  
  engineer: [
    Permission.ASSIGN_PERSONNEL_SKILLS,
    Permission.MANAGE_EQUIPMENT,
    Permission.MANAGE_METHODS,
    Permission.MANAGE_MATERIALS,
    Permission.MANAGE_SHIFTS,
    Permission.CREATE_WORK_ORDER,
    Permission.CREATE_SUBTASK,
    Permission.ASSIGN_TECHNICIAN,
    Permission.EXECUTE_SUBTASK,
    Permission.VERIFY_RESULTS,
    Permission.VIEW_WORK_ORDER_QUERY,
    Permission.INITIATE_HANDOVER,
    Permission.ALLOCATE_MATERIALS,
    Permission.HANDLE_MATERIAL_RETURN,
    Permission.CONFIRM_MATERIAL_ALERTS,
    Permission.VIEW_LAB_DASHBOARD,
    Permission.VIEW_CYCLE_TIME_REPORT,
    Permission.VIEW_SKILLS_MATRIX,
    Permission.VIEW_REPORTS,
  ],
  
  technician: [
    Permission.EXECUTE_SUBTASK,
    Permission.VIEW_WORK_ORDER_QUERY,
    Permission.INITIATE_HANDOVER,
    Permission.HANDLE_MATERIAL_RETURN,
    Permission.CONFIRM_MATERIAL_ALERTS,
    Permission.VIEW_LAB_DASHBOARD,
    Permission.VIEW_CYCLE_TIME_REPORT,
    Permission.VIEW_SKILLS_MATRIX,
    Permission.VIEW_REPORTS,
  ],
  
  viewer: [
    Permission.VIEW_WORK_ORDER_QUERY, // Viewer can only access work order query
  ],
};

// Route to required permissions mapping
export const routePermissions: Record<string, Permission[]> = {
  '/dashboard': [Permission.VIEW_LAB_DASHBOARD],
  '/equipment-dashboard': [Permission.VIEW_LAB_DASHBOARD],
  '/personnel-dashboard': [Permission.VIEW_LAB_DASHBOARD],
  '/locations': [Permission.MANAGE_LOCATIONS],
  '/sites': [Permission.MANAGE_SITES],
  '/laboratories': [Permission.MANAGE_LABORATORIES],
  '/personnel': [Permission.MANAGE_SHIFTS], // 需要人员管理权限
  '/skills': [Permission.VIEW_SKILLS_MATRIX],
  '/skills-config': [Permission.MANAGE_SKILLS],
  '/transfers': [Permission.INITIATE_BORROW],
  '/shifts': [Permission.MANAGE_SHIFTS],
  '/equipment': [Permission.MANAGE_EQUIPMENT],
  '/methods': [Permission.MANAGE_METHODS],
  '/materials': [Permission.MANAGE_MATERIALS],
  '/work-orders': [Permission.VIEW_LAB_DASHBOARD], // All can view, actions controlled separately
  '/work-order-query': [Permission.VIEW_WORK_ORDER_QUERY], // Viewer-accessible
  '/clients': [Permission.MANAGE_CLIENTS],
  '/client-slas': [Permission.MANAGE_CLIENT_SLA],
  '/source-categories': [Permission.MANAGE_SOURCE_CATEGORIES],
  '/products': [Permission.MANAGE_CLIENTS], // 产品管理需要客户管理权限
  '/handovers': [Permission.INITIATE_HANDOVER],
  '/audit-logs': [Permission.VIEW_AUDIT_LOGS],
  '/user-management': [Permission.MANAGE_USERS], // Admin only
  '/settings': [Permission.MANAGE_USERS], // 仅管理员可访问系统设置页面
};

// Check if user has a specific permission
export function hasPermission(role: UserRole | undefined, permission: Permission): boolean {
  if (!role) return false;
  return rolePermissions[role]?.includes(permission) ?? false;
}

// Check if user has any of the specified permissions
export function hasAnyPermission(role: UserRole | undefined, permissions: Permission[]): boolean {
  if (!role) return false;
  return permissions.some(p => hasPermission(role, p));
}

// Check if user has all of the specified permissions
export function hasAllPermissions(role: UserRole | undefined, permissions: Permission[]): boolean {
  if (!role) return false;
  return permissions.every(p => hasPermission(role, p));
}

// Check if user can access a specific route
export function canAccessRoute(role: UserRole | undefined, route: string): boolean {
  if (!role) return false;
  
  // Admin can access everything
  if (role === 'admin') return true;
  
  // Get required permissions for the route
  const requiredPermissions = routePermissions[route];
  
  // If no permissions defined, allow access
  if (!requiredPermissions || requiredPermissions.length === 0) return true;
  
  // Check if user has any of the required permissions
  return hasAnyPermission(role, requiredPermissions);
}

// Get all permissions for a role
export function getRolePermissions(role: UserRole | undefined): Permission[] {
  if (!role) return [];
  return rolePermissions[role] || [];
}

// Get readable permission names (Chinese)
export const permissionLabels: Record<Permission, string> = {
  [Permission.MANAGE_USERS]: '管理用户',
  [Permission.MANAGE_SITES]: '管理站点',
  [Permission.MANAGE_LABORATORIES]: '管理实验室',
  [Permission.MANAGE_LOCATIONS]: '管理位置',
  [Permission.MANAGE_CLIENTS]: '管理客户',
  [Permission.MANAGE_CLIENT_SLA]: '管理客户SLA',
  [Permission.MANAGE_SOURCE_CATEGORIES]: '管理测试来源类别',
  [Permission.MANAGE_SKILL_CATEGORIES]: '管理技能分类',
  [Permission.MANAGE_SKILLS]: '管理技能',
  [Permission.ASSIGN_PERSONNEL_SKILLS]: '分配人员技能',
  [Permission.INITIATE_BORROW]: '发起人员借调',
  [Permission.APPROVE_BORROW]: '审批人员借调',
  [Permission.MANAGE_SHIFTS]: '管理班次',
  [Permission.MANAGE_EQUIPMENT]: '管理设备',
  [Permission.MANAGE_METHODS]: '管理分析/测试方法',
  [Permission.CREATE_WORK_ORDER]: '创建工单',
  [Permission.ASSIGN_LEAD_ENGINEER]: '指派主管工程师',
  [Permission.CREATE_SUBTASK]: '创建子任务',
  [Permission.ASSIGN_TECHNICIAN]: '分配技术员',
  [Permission.EXECUTE_SUBTASK]: '执行子任务',
  [Permission.VERIFY_RESULTS]: '验收签核结果',
  [Permission.VIEW_WORK_ORDER_QUERY]: '查看工单查询',
  [Permission.INITIATE_HANDOVER]: '发起交接',
  [Permission.MANAGE_MATERIALS]: '管理材料',
  [Permission.ALLOCATE_MATERIALS]: '分配材料到任务',
  [Permission.HANDLE_MATERIAL_RETURN]: '处理材料归还/报废',
  [Permission.CONFIRM_MATERIAL_ALERTS]: '确认材料提醒',
  [Permission.VIEW_LAB_DASHBOARD]: '查看本实验室仪表板',
  [Permission.VIEW_ALL_DASHBOARDS]: '查看所有实验室仪表板',
  [Permission.VIEW_CYCLE_TIME_REPORT]: '查看周期时间报告',
  [Permission.VIEW_SKILLS_MATRIX]: '查看技能矩阵',
  [Permission.VIEW_REPORTS]: '查看报表',
  [Permission.VIEW_AUDIT_LOGS]: '查看审计日志',
};

// Permission categories for display
export const permissionCategories: Record<string, { label: string; permissions: Permission[] }> = {
  system: {
    label: '系统管理',
    permissions: [
      Permission.MANAGE_USERS,
      Permission.MANAGE_SITES,
      Permission.MANAGE_LABORATORIES,
      Permission.MANAGE_LOCATIONS,
    ],
  },
  clients: {
    label: '客户与SLA',
    permissions: [
      Permission.MANAGE_CLIENTS,
      Permission.MANAGE_CLIENT_SLA,
      Permission.MANAGE_SOURCE_CATEGORIES,
    ],
  },
  skills: {
    label: '技能管理',
    permissions: [
      Permission.MANAGE_SKILL_CATEGORIES,
      Permission.MANAGE_SKILLS,
      Permission.ASSIGN_PERSONNEL_SKILLS,
    ],
  },
  personnel: {
    label: '人员管理',
    permissions: [
      Permission.INITIATE_BORROW,
      Permission.APPROVE_BORROW,
      Permission.MANAGE_SHIFTS,
    ],
  },
  equipment: {
    label: '设备与方法',
    permissions: [
      Permission.MANAGE_EQUIPMENT,
      Permission.MANAGE_METHODS,
    ],
  },
  workOrders: {
    label: '工单管理',
    permissions: [
      Permission.CREATE_WORK_ORDER,
      Permission.ASSIGN_LEAD_ENGINEER,
      Permission.CREATE_SUBTASK,
      Permission.ASSIGN_TECHNICIAN,
      Permission.EXECUTE_SUBTASK,
      Permission.VERIFY_RESULTS,
      Permission.VIEW_WORK_ORDER_QUERY,
      Permission.INITIATE_HANDOVER,
    ],
  },
  materials: {
    label: '材料管理',
    permissions: [
      Permission.MANAGE_MATERIALS,
      Permission.ALLOCATE_MATERIALS,
      Permission.HANDLE_MATERIAL_RETURN,
      Permission.CONFIRM_MATERIAL_ALERTS,
    ],
  },
  reports: {
    label: '仪表板与报表',
    permissions: [
      Permission.VIEW_LAB_DASHBOARD,
      Permission.VIEW_ALL_DASHBOARDS,
      Permission.VIEW_CYCLE_TIME_REPORT,
      Permission.VIEW_SKILLS_MATRIX,
      Permission.VIEW_REPORTS,
      Permission.VIEW_AUDIT_LOGS,
    ],
  },
};

// 角色默认首页映射
// Default home page for each role based on their primary permissions
const roleHomePages: Record<UserRole, string> = {
  admin: '/dashboard',           // 管理员 - 仪表板
  manager: '/dashboard',         // 经理 - 仪表板
  engineer: '/dashboard',        // 工程师 - 仪表板
  technician: '/dashboard',      // 技术员 - 仪表板
  viewer: '/work-order-query',   // 查看者 - 工单查询 (唯一可访问页面)
};

// 获取角色对应的首页路径
// Get the appropriate home page for a user role
export function getRoleHomePage(role: UserRole | undefined): string {
  if (!role) return '/login';
  return roleHomePages[role] || '/work-order-query';
}

// 获取用户第一个可访问的路由 (用于403页面导航)
// Get the first accessible route for a user (for 403 page navigation)
export function getFirstAccessibleRoute(role: UserRole | undefined): string {
  if (!role) return '/login';
  
  // 首先尝试返回角色默认首页
  const homePage = roleHomePages[role];
  if (canAccessRoute(role, homePage)) {
    return homePage;
  }
  
  // 如果默认首页不可访问，查找第一个可访问的路由
  const routeOrder = [
    '/work-order-query',
    '/dashboard',
    '/work-orders',
    '/personnel',
    '/equipment',
    '/materials',
    '/settings',
  ];
  
  for (const route of routeOrder) {
    if (canAccessRoute(role, route)) {
      return route;
    }
  }
  
  return '/login';
}

// ============================================================================
// Module Permission Functions
// ============================================================================

// 用户可访问的模块列表（从后端获取后存储在此）
let userAccessibleModules: ModuleCode[] = [];

// 设置用户可访问的模块列表
export function setUserAccessibleModules(modules: ModuleCode[]): void {
  userAccessibleModules = modules;
}

// 获取用户可访问的模块列表
export function getUserAccessibleModules(): ModuleCode[] {
  return userAccessibleModules;
}

// 清除用户模块权限（登出时调用）
export function clearUserModulePermissions(): void {
  userAccessibleModules = [];
}

// 检查用户是否可以访问指定模块
export function canAccessModule(moduleCode: ModuleCode): boolean {
  // 如果还没有加载模块权限，默认允许访问
  if (userAccessibleModules.length === 0) {
    return true;
  }
  return userAccessibleModules.includes(moduleCode);
}

// 检查用户是否可以访问指定路由（基于模块权限）
export function canAccessRouteByModule(route: string): boolean {
  const moduleCode = routeToModule[route];
  
  // 如果路由没有映射到模块，允许访问
  if (!moduleCode) {
    return true;
  }
  
  return canAccessModule(moduleCode);
}

// 综合检查用户是否可以访问路由（同时检查操作权限和模块权限）
export function canAccessRouteWithModuleCheck(role: UserRole | undefined, route: string): boolean {
  // 首先检查操作权限
  if (!canAccessRoute(role, route)) {
    return false;
  }
  
  // 然后检查模块权限
  return canAccessRouteByModule(route);
}

File: backend/app/api/v1/endpoints/permissions.py
"""
Permission Management API endpoints - Admin only.
Provides CRUD operations for dynamic role permissions.
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import Optional

from app.core.database import get_db
from app.schemas.permission import (
    RolePermissionUpdate, RolePermissionResponse, PermissionMatrixResponse,
    RolePermissionsResponse, PermissionMatrixItem, PermissionChangeLogResponse,
    BulkPermissionUpdate, PermissionDefinition
)
from app.api.deps import require_admin
from app.models.user import User, UserRole
from app.models.permission import RolePermission, PermissionChangeLog

router = APIRouter(prefix="/permissions", tags=["Permission Management"])

# Permission labels in Chinese
PERMISSION_LABELS = {
    "manage_users": "管理用户",
    "manage_sites": "管理站点",
    "manage_laboratories": "管理实验室",
    "manage_locations": "管理位置",
    "manage_clients": "管理客户",
    "manage_client_sla": "管理客户SLA",
    "manage_source_categories": "管理测试来源类别",
    "manage_skill_categories": "管理技能分类",
    "manage_skills": "管理技能",
    "assign_personnel_skills": "分配人员技能",
    "initiate_borrow": "发起人员借调",
    "approve_borrow": "审批人员借调",
    "manage_shifts": "管理班次",
    "manage_equipment": "管理设备",
    "manage_methods": "管理分析/测试方法",
    "create_work_order": "创建工单",
    "assign_lead_engineer": "指派主管工程师",
    "create_subtask": "创建子任务",
    "assign_technician": "分配技术员",
    "execute_subtask": "执行子任务",
    "verify_results": "验收签核结果",
    "initiate_handover": "发起交接",
    "manage_materials": "管理材料",
    "allocate_materials": "分配材料到任务",
    "handle_material_return": "处理材料归还/报废",
    "confirm_material_alerts": "确认材料提醒",
    "view_lab_dashboard": "查看本实验室仪表板",
    "view_all_dashboards": "查看所有实验室仪表板",
    "view_cycle_time_report": "查看周期时间报告",
    "view_skills_matrix": "查看技能矩阵",
    "view_reports": "查看报表",
    "view_audit_logs": "查看审计日志",
    "view_work_order_query": "查看工单查询",  # Viewer-specific permission
}

# Role labels in Chinese
ROLE_LABELS = {
    "admin": "系统管理员",
    "manager": "经理",
    "engineer": "工程师",
    "technician": "技术员",
    "viewer": "访客",
}

# Permission categories
PERMISSION_CATEGORIES = {
    "system": {
        "label": "系统管理",
        "permissions": ["manage_users", "manage_sites", "manage_laboratories", "manage_locations"],
    },
    "clients": {
        "label": "客户与SLA",
        "permissions": ["manage_clients", "manage_client_sla", "manage_source_categories"],
    },
    "skills": {
        "label": "技能管理",
        "permissions": ["manage_skill_categories", "manage_skills", "assign_personnel_skills"],
    },
    "personnel": {
        "label": "人员管理",
        "permissions": ["initiate_borrow", "approve_borrow", "manage_shifts"],
    },
    "equipment": {
        "label": "设备与方法",
        "permissions": ["manage_equipment", "manage_methods"],
    },
    "work_orders": {
        "label": "工单管理",
        "permissions": [
            "create_work_order", "assign_lead_engineer", "create_subtask",
            "assign_technician", "execute_subtask", "verify_results", "initiate_handover",
            "view_work_order_query"
        ],
    },
    "materials": {
        "label": "材料管理",
        "permissions": ["manage_materials", "allocate_materials", "handle_material_return", "confirm_material_alerts"],
    },
    "reports": {
        "label": "仪表板与报表",
        "permissions": [
            "view_lab_dashboard", "view_all_dashboards", "view_cycle_time_report",
            "view_skills_matrix", "view_reports", "view_audit_logs"
        ],
    },
}

# Default permissions for each role
DEFAULT_ROLE_PERMISSIONS = {
    "admin": list(PERMISSION_LABELS.keys()),  # Admin has all permissions
    "manager": [
        "manage_locations", "manage_clients", "manage_client_sla", "manage_source_categories",
        "manage_skill_categories", "manage_skills", "assign_personnel_skills",
        "initiate_borrow", "approve_borrow", "manage_shifts",
        "manage_equipment", "manage_methods",
        "create_work_order", "assign_lead_engineer", "create_subtask", "assign_technician",
        "execute_subtask", "verify_results", "initiate_handover", "view_work_order_query",
        "manage_materials", "allocate_materials", "handle_material_return", "confirm_material_alerts",
        "view_lab_dashboard", "view_all_dashboards", "view_cycle_time_report",
        "view_skills_matrix", "view_reports", "view_audit_logs",
    ],
    "engineer": [
        "assign_personnel_skills", "manage_methods",
        "create_work_order", "create_subtask", "assign_technician",
        "execute_subtask", "verify_results", "initiate_handover", "view_work_order_query",
        "allocate_materials", "handle_material_return", "confirm_material_alerts",
        "view_lab_dashboard", "view_cycle_time_report", "view_skills_matrix", "view_reports",
    ],
    "technician": [
        "execute_subtask", "initiate_handover", "view_work_order_query",
        "handle_material_return", "confirm_material_alerts",
        "view_lab_dashboard", "view_cycle_time_report", "view_skills_matrix", "view_reports",
    ],
    "viewer": [
        "view_work_order_query",  # Viewer can only access work order query
    ],
}


def get_permission_category(permission: str) -> tuple[str, str]:
    """Get category code and label for a permission."""
    for cat_code, cat_info in PERMISSION_CATEGORIES.items():
        if permission in cat_info["permissions"]:
            return cat_code, cat_info["label"]
    return "other", "其他"


def initialize_default_permissions(db: Session):
    """Initialize default permissions if not already set."""
    for role, permissions in DEFAULT_ROLE_PERMISSIONS.items():
        for permission in PERMISSION_LABELS.keys():
            existing = db.query(RolePermission).filter(
                RolePermission.role == role,
                RolePermission.permission == permission
            ).first()
            
            if not existing:
                is_enabled = permission in permissions
                db.add(RolePermission(
                    role=role,
                    permission=permission,
                    is_enabled=is_enabled
                ))
    
    db.commit()


@router.get("/definitions", response_model=list[PermissionDefinition])
def get_permission_definitions(
    _: User = Depends(require_admin),
):
    """Get all permission definitions with labels and categories."""
    definitions = []
    for permission, label in PERMISSION_LABELS.items():
        cat_code, cat_label = get_permission_category(permission)
        definitions.append(PermissionDefinition(
            code=permission,
            label=label,
            category=cat_code,
            category_label=cat_label
        ))
    return definitions


@router.get("/matrix", response_model=PermissionMatrixResponse)
def get_permission_matrix(
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    """Get the complete permission matrix for all roles."""
    # Initialize defaults if needed
    initialize_default_permissions(db)
    
    roles_data = []
    for role in ["admin", "manager", "engineer", "technician", "viewer"]:
        permissions_list = []
        
        for permission in PERMISSION_LABELS.keys():
            role_perm = db.query(RolePermission).filter(
                RolePermission.role == role,
                RolePermission.permission == permission
            ).first()
            
            # For admin, always show as enabled
            if role == "admin":
                is_enabled = True
            else:
                is_enabled = role_perm.is_enabled if role_perm else False
            
            permissions_list.append(PermissionMatrixItem(
                permission=permission,
                permission_label=PERMISSION_LABELS[permission],
                is_enabled=is_enabled
            ))
        
        roles_data.append(RolePermissionsResponse(
            role=role,
            role_label=ROLE_LABELS[role],
            permissions=permissions_list
        ))
    
    return PermissionMatrixResponse(roles=roles_data)


@router.get("/role/{role}", response_model=RolePermissionsResponse)
def get_role_permissions(
    role: str,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    """Get all permissions for a specific role."""
    if role not in ROLE_LABELS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid role: {role}"
        )
    
    # Initialize defaults if needed
    initialize_default_permissions(db)
    
    permissions_list = []
    for permission in PERMISSION_LABELS.keys():
        role_perm = db.query(RolePermission).filter(
            RolePermission.role == role,
            RolePermission.permission == permission
        ).first()
        
        # For admin, always show as enabled
        if role == "admin":
            is_enabled = True
        else:
            is_enabled = role_perm.is_enabled if role_perm else False
        
        permissions_list.append(PermissionMatrixItem(
            permission=permission,
            permission_label=PERMISSION_LABELS[permission],
            is_enabled=is_enabled
        ))
    
    return RolePermissionsResponse(
        role=role,
        role_label=ROLE_LABELS[role],
        permissions=permissions_list
    )


@router.put("/role/{role}/{permission}")
def update_role_permission(
    role: str,
    permission: str,
    update_data: RolePermissionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Update a specific permission for a role."""
    # Validate role
    if role not in ROLE_LABELS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid role: {role}"
        )
    
    # Cannot modify admin permissions
    if role == "admin":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot modify admin permissions"
        )
    
    # Validate permission
    if permission not in PERMISSION_LABELS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid permission: {permission}"
        )
    
    # Get or create role permission
    role_perm = db.query(RolePermission).filter(
        RolePermission.role == role,
        RolePermission.permission == permission
    ).first()
    
    old_value = role_perm.is_enabled if role_perm else None
    
    if role_perm:
        role_perm.is_enabled = update_data.is_enabled
    else:
        role_perm = RolePermission(
            role=role,
            permission=permission,
            is_enabled=update_data.is_enabled
        )
        db.add(role_perm)
    
    # Log the change
    change_log = PermissionChangeLog(
        role=role,
        permission=permission,
        old_value=old_value,
        new_value=update_data.is_enabled,
        changed_by_id=current_user.id,
        reason=update_data.reason
    )
    db.add(change_log)
    
    db.commit()
    
    return {
        "message": "Permission updated successfully",
        "role": role,
        "permission": permission,
        "is_enabled": update_data.is_enabled
    }


@router.post("/bulk-update")
def bulk_update_permissions(
    update_data: BulkPermissionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Bulk update multiple permissions."""
    updated_count = 0
    
    for item in update_data.updates:
        role = item.get("role")
        permission = item.get("permission")
        is_enabled = item.get("is_enabled")
        
        # Skip invalid entries
        if not role or not permission or is_enabled is None:
            continue
        
        # Skip admin role
        if role == "admin":
            continue
        
        # Skip invalid roles/permissions
        if role not in ROLE_LABELS or permission not in PERMISSION_LABELS:
            continue
        
        # Get or create role permission
        role_perm = db.query(RolePermission).filter(
            RolePermission.role == role,
            RolePermission.permission == permission
        ).first()
        
        old_value = role_perm.is_enabled if role_perm else None
        
        if role_perm:
            if role_perm.is_enabled != is_enabled:
                role_perm.is_enabled = is_enabled
                updated_count += 1
                
                # Log the change
                db.add(PermissionChangeLog(
                    role=role,
                    permission=permission,
                    old_value=old_value,
                    new_value=is_enabled,
                    changed_by_id=current_user.id,
                    reason=update_data.reason
                ))
        else:
            db.add(RolePermission(
                role=role,
                permission=permission,
                is_enabled=is_enabled
            ))
            updated_count += 1
            
            # Log the change
            db.add(PermissionChangeLog(
                role=role,
                permission=permission,
                old_value=None,
                new_value=is_enabled,
                changed_by_id=current_user.id,
                reason=update_data.reason
            ))
    
    db.commit()
    
    return {
        "message": f"Updated {updated_count} permissions",
        "updated_count": updated_count
    }


@router.get("/change-logs", response_model=list[PermissionChangeLogResponse])
def get_permission_change_logs(
    role: Optional[str] = Query(None),
    permission: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    """Get permission change history logs."""
    query = db.query(PermissionChangeLog).order_by(PermissionChangeLog.changed_at.desc())
    
    if role:
        query = query.filter(PermissionChangeLog.role == role)
    
    if permission:
        query = query.filter(PermissionChangeLog.permission == permission)
    
    logs = query.limit(limit).all()
    
    return [PermissionChangeLogResponse.model_validate(log) for log in logs]


@router.post("/reset-to-defaults")
def reset_permissions_to_defaults(
    role: Optional[str] = Query(None, description="Reset specific role or all if not specified"),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    """Reset permissions to default values."""
    roles_to_reset = [role] if role else list(ROLE_LABELS.keys())
    
    # Skip admin
    roles_to_reset = [r for r in roles_to_reset if r != "admin"]
    
    reset_count = 0
    
    for r in roles_to_reset:
        if r not in DEFAULT_ROLE_PERMISSIONS:
            continue
        
        default_perms = DEFAULT_ROLE_PERMISSIONS[r]
        
        for permission in PERMISSION_LABELS.keys():
            should_be_enabled = permission in default_perms
            
            role_perm = db.query(RolePermission).filter(
                RolePermission.role == r,
                RolePermission.permission == permission
            ).first()
            
            if role_perm:
                if role_perm.is_enabled != should_be_enabled:
                    old_value = role_perm.is_enabled
                    role_perm.is_enabled = should_be_enabled
                    reset_count += 1
                    
                    db.add(PermissionChangeLog(
                        role=r,
                        permission=permission,
                        old_value=old_value,
                        new_value=should_be_enabled,
                        changed_by_id=current_user.id,
                        reason="Reset to default"
                    ))
            else:
                db.add(RolePermission(
                    role=r,
                    permission=permission,
                    is_enabled=should_be_enabled
                ))
                reset_count += 1
    
    db.commit()
    
    return {
        "message": f"Reset {reset_count} permissions to defaults",
        "reset_count": reset_count,
        "roles_affected": roles_to_reset
    }


@router.get("/user/{user_id}/effective")
def get_user_effective_permissions(
    user_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
):
    """Get effective permissions for a specific user based on their role."""
    from app.models.user import User as UserModel
    
    user = db.query(UserModel).filter(UserModel.id == user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    role = user.role.value if hasattr(user.role, 'value') else str(user.role)
    
    # Initialize defaults if needed
    initialize_default_permissions(db)
    
    effective_permissions = []
    
    for permission in PERMISSION_LABELS.keys():
        if role == "admin":
            is_enabled = True
        else:
            role_perm = db.query(RolePermission).filter(
                RolePermission.role == role,
                RolePermission.permission == permission
            ).first()
            is_enabled = role_perm.is_enabled if role_perm else False
        
        if is_enabled:
            effective_permissions.append(permission)
    
    return {
        "user_id": user_id,
        "username": user.username,
        "role": role,
        "role_label": ROLE_LABELS.get(role, role),
        "permissions": effective_permissions
    }
